### ISSUE TITLE
generating compile_commands.json for workspace

### ISSUE THREAD

**Issue by rhklite** [catkin/catkin_tools#551]
[//]: # (Thank you for reporting an issue for catkin_tools!)

[//]: # (Please fill out the details below so that your issue can be resolved quickly!)
[//]: # (If you want to suggest a feature or start a discussion, feel free to delete them.)

### System Info

* Operating System: `uname -a`
* Python Version: `python --version`
* Version of catkin_tools: `catkin --version` OR *Git revision*
* ROS Distro: `echo $ROS_DISTRO`

### Build / Run Issue

[//]: # (If you are migrating from catkin_make, please follow the migration guide:)
[//]: # (http://catkin-tools.readthedocs.org/en/latest/migration.html)

[//]: # (Please also check for solved issues here:)
[//]: # (http://catkin-tools.readthedocs.org/en/latest/troubleshooting.html)

[//]: # (And check for open issues here:)
[//]: # (https://github.com/catkin/catkin_tools/issues?q=is%3Aopen+is%3Aissue+label%3Abug)

[//]: # (If you have an issue building and running code, please verify the following:)

* [x] Works with `catkin_make`
* [ ] Works with `catkin_make_isolated --merge`
* [ ] Works with `catkin build`
* [ ] Works with `catkin build -p1`
* [ ] I did  not `read this`

### Expected Behavior

[//]: # (Please describe what you expected or think should happen.)
- `catkin build -DCMAKE_EXPORT_COMPILE_COMMANDS=1` should generate a `compile_commands.json` file inside `catkin_ws/build` that's for the entire workspace. I'm using this file for the cquery linter

### Actual Behavior

- I'm trying to get a `compile_commands.json` output with `catkin build`, the same file that you would get when you use `cakin_make`
- **catkin_make**: `catkin_make -DCMAKE_EXPORT_COMPILE_COMMANDS=1` would generate a `compile_commands.json` inside `catkin_ws/build` which is a nice file that works for the entire workspace.
- **catkin build**: `catkin build -DCMAKE_EXPORT_COMPILE_COMMANDS=1` doesn't generate a nice file for the entire workspace, but rather individual `compile_commands.json` for each package. I tried to do a work around by using a python file to concatenate all the `compile_commands.json` in each subfolder in the `build` folder. However the generated `compile_commands.json` doesn't recognise `#include <message_header>.h`, where as the `compile_commands.json` generated through catkin_make does.

Is there a way to properly generate a `compile_commands.json` for the entire catkin workspace using `catkin build`?

python script here:
```python
#!/usr/bin/env python

import os

# create a dictionary with file names as keys
# and for each file name the paths where they
# were found
file_paths = {}
for root, dirs, files in os.walk('.'):
    for f in files:
        if f.endswith('compile_commands.json'):
            if f not in file_paths:
                file_paths[f] = []
            file_paths[f].append(root)

# for each file in the dictionary, concatenate
# the content of the files in each directory
# and write the merged content into a file
# with the same name at the top directory
for f, paths in file_paths.items():
    txt = []
    for p in paths:
        with open(os.path.join(p, f)) as f2:
            txt.append(f2.read())
    with open(f, 'w') as f3:
        f3.write(''.join(txt))
```

[//]: # (Please describe what actually happens.)

[//]: # (Please put long CLI output into Gists: http://gist.github.com)
[//]: # (Use `VERBOSE=1 catkin build -p1 -v` for verbose output)
[//]: # (Please post `logs` directory if available.)

### Steps to Reproduce the Issue

[//]: # (Please give a series of command-line commands which cause the problem.)
[//]: # (Include a checkout URI for the workspace so that we can reproduce the issue.)
[//]: # (If we can't reproduce it on our own machines, we can't fix it!)

```bash
cd ~/catkin_ws
catkin clean
catkin build -DCMAKE_EXPORT_COMPILE_COMMANDS=1
```

**Comment by rhklite** [comment]
there was a bug in my work around python script, i made edits to the script and got it working.

Is there a better way to do this? to generate a `compile_commands.json` without needing a pythonc script?

**Comment by ndepal** [comment]
I agree, it would be great if `catkin build -DCMAKE_EXPORT_COMPILE_COMMANDS=1` did this.

Based on your python script I made my own shell script that does the same:

```sh
#!/bin/sh

cd `catkin locate --workspace $(pwd)`

concatenated="build/compile_commands.json"

echo "[" > $concatenated

first=1
for d in build/*
do
    f="$d/compile_commands.json"

    if test -f "$f"; then
        if [ $first -eq 0 ]; then
            echo "," >> $concatenated
        fi

        cat $f | sed '1d;$d' >> $concatenated
    fi

    first=0
done

echo "]" >> $concatenated
```

**Comment by kgreenek** [comment]
If you happen to use YouCompleteMe with vim, you can use this plugin so you don't have to generate the merged compile_commands.json file. Just run `catkin config --cmake-args='-DCMAKE_EXPORT_COMPILE_COMMANDS=ON'`, install the plugin to vim, and everything will just work when you're inside the catkin workspace. Alternatively, you can place the .ycm_extra_conf.py file form that repo in the root of your catkin workspace.

https://github.com/kgreenek/vim-ros-ycm

**Comment by LeroyR** [comment]
It generates multiple compile commands instead of one merged file as you are working with multiple projects inside. Merging the compile commands sounds like breaking the separation between multiple projects.

So while i understand why you/we/i would want this, it does not sound like something catkin tools should ever do.

**Comment by AlexisTM** [comment]
@ndepal @rhklite 

My way to merge all compile_commands.json together, using the jq tool to simplify integration with other tools.

```bash
sudo apt install jq

jq -s 'map(.[])' build_folder/**/compile_commands.json > compile_commands.json
```

@LeroyR I agree, catkin_tools should keep compile commands split as it is a split workflow.

**Comment by hgaiser** [comment]
I am doing something similar. It isn't pretty, but it doesn't require an extra tool:

```
printf '[' > compile_commands.json
find ./build_isolated -type f -name 'compile_commands.json' -exec sh -c "cat {} | tail -n+2 | head -n-1 && printf ','" >> compile_commands.json \;
sed -i '$s/.$//' compile_commands.json
printf '\n]\n' >> compile_commands.json
```

**Comment by stertingen** [comment]
Interestingly, colcon does this: https://github.com/colcon/colcon-cmake/pull/69

**Comment by zoenglinghou** [comment]
I used @AlexisTM 's comment and come up with a [script](https://github.com/zoenglinghou/catkin-compile-commands)

**Comment by wafts95** [comment]
I came across another use case for having a common compile_commands.json file, and that is to enable the use of the SciTools Understand IDE [(link)](https://www.scitools.com/) with ROS projects.
